src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -209,48 +211,201 @@ internal static IEnumerable<string> GetCommaSeparatedValues(string text)
+                var descriptor = ToAttributeDescriptor(property, attributeNameAttribute);
                 if (ValidateTagHelperAttributeDescriptor(descriptor, type, errorSink))
                 {
+                    // Does this property also support indexer assignments?
+                    bool isInvalid;
+                    var indexerDescriptor = ToIndexerAttributeDescriptor(
+                        property,
+                        attributeNameAttribute,
+                        parentType: type,
+                        errorSink: errorSink,
+                        defaultPrefix: descriptor.Name + "-",
+                        isInvalid: out isInvalid);
+
+                    if (indexerDescriptor != null &&
+                        !ValidateTagHelperAttributeDescriptor(indexerDescriptor, type, errorSink))
+                    {
+                        isInvalid = true;
+                    }
+
+                    if (isInvalid)
+                    {
+                        // HtmlAttributeNameAttribute was not valid. Ignore this property completely.
+                        continue;
+                    }
+
                     attributeDescriptors.Add(descriptor);
+                    if (indexerDescriptor != null)
+                    {
+                        indexerDescriptors.Add(indexerDescriptor);
+                    }
                 }
             }
 
+            attributeDescriptors.AddRange(indexerDescriptors);
+
             return attributeDescriptors;
         }
 
-        private static bool ValidateTagHelperAttributeDescriptor(
+        // Internal for testing.
+        internal static bool ValidateTagHelperAttributeDescriptor(
             TagHelperAttributeDescriptor attributeDescriptor,
             Type parentType,
             ErrorSink errorSink)
         {
+            if (attributeDescriptor == null)
+            {
+                return false;
+            }
+
+            var nameOrPrefix = attributeDescriptor.IsIndexer ?
+                Resources.TagHelperDescriptorFactory_Prefix :
+                Resources.TagHelperDescriptorFactory_Name;
+            return ValidateTagHelperAttributeNameOrPrefix(
+                attributeDescriptor.Name,
+                parentType,
+                attributeDescriptor.PropertyName,
+                errorSink,
+                nameOrPrefix);
+        }
+
+        private static bool ValidateTagHelperAttributeNameOrPrefix(
+            string attributeNameOrPrefix,
+            Type parentType,
+            string propertyName,
+            ErrorSink errorSink,
+            string nameOrPrefix)
+        {
+            if (string.IsNullOrEmpty(attributeNameOrPrefix))
+            {
+                // HtmlAttributeNameAttribute validates Name is non-null and non-empty. Both are valid for
+                // DictionaryAttributePrefix. (Empty DictionaryAttributePrefix is a corner case which would bind every
```

Comment:
Ah bah, I see below that you handled this error case, should remove that part of this comment.